ugc/storage.cpp

```diff
@@ -1,33 +1,434 @@
+    return {};
+
+  auto const offset = index->m_offset;
+  auto const nextIndex = index + 1;
+  uint64_t nextOffset;
+  if (nextIndex == m_UGCIndexes.end())
+    CHECK(GetPlatform().GetFileSizeByName(kUGCUpdateFileName, nextOffset), ());
+  else
+    nextOffset = nextIndex->m_offset;
+
+  auto const size = nextOffset - offset;
+  vector<uint8_t> buf;
+  auto const ugcFilePath = GetUGCFilePath();
+  try
+  {
+    FileReader r(ugcFilePath);
+    r.Read(offset, buf.data(), size);
+  }
+  catch (RootException const & exception)
+  {
+    LOG(LWARNING, ("Exception while reading file:", ugcFilePath));
+    return {};
+  }
+
+  MemReader r(buf.data(), buf.size());
+  NonOwningReaderSource source(r);
+  UGCUpdate update;
+  Deserialize(source, update);
+  return update;
 }
 
 void Storage::SetUGCUpdate(FeatureID const & id, UGCUpdate const & ugc)
 {
-  m_ugc[id] = ugc;
+  ASSERT_THREAD_CHECKER(m_threadChecker, ());
+  auto const feature = GetOriginalFeature(id);
+  CHECK(feature, ());
+
+  auto const & mercator = feature->GetCenter();
+  feature::TypesHolder th(*feature);
+  th.SortBySpec();
+  auto const type = th.GetBestType();
+  for (auto & index : m_UGCIndexes)
+  {
+    if (type == index.m_type && mercator == index.m_mercator && !index.m_isDeleted)
+    {
+      index.m_isDeleted = true;
+      m_numberOfDeleted++;
+      break;
+    }
+  }
+  // TODO: Call Defragmentation().
+
+  auto const ugcFilePath = GetUGCFilePath();
+  try
+  {
+    FileWriter w(ugcFilePath);
+    Serialize(w, ugc);
+  }
+  catch (RootException const & exception)
+  {
+    LOG(LWARNING, ("Exception while writing file:", ugcFilePath));
+    return;
+  }
+
+  UGCIndex index;
+  index.m_mercator = mercator;
+  uint64_t offset;
+  if (GetUGCFileSize(offset))
+    offset = 0;
+
+  index.m_type = type;
+  index.m_mwmName = id.GetMwmName();
+  index.m_dataVersion = id.GetMwmVersion();
+  index.m_featureId = id.m_index;
+  index.m_offset = offset;
+  m_UGCIndexes.emplace_back(move(index));
 }
 
 void Storage::Load()
 {
+  ASSERT_THREAD_CHECKER(m_threadChecker, ());
+  string data;
+  auto const indexFilePath = GetIndexFilePath();
+  try
+  {
+    FileReader r(indexFilePath);
+    r.ReadAsString(data);
+  }
+  catch (RootException const & exception)
+  {
+    LOG(LWARNING, ("Exception while reading file:", indexFilePath));
+    return;
+  }
+
+  DeserializeUGCIndex(data, m_UGCIndexes, m_numberOfDeleted);
+  sort(m_UGCIndexes.begin(), m_UGCIndexes.end(), [](UGCIndex const & lhs, UGCIndex const & rhs) -> bool {
+    return lhs.m_offset < rhs.m_offset;
+  });
+}
+
+void Storage::SaveIndex() const
+{
+  ASSERT_THREAD_CHECKER(m_threadChecker, ());
+  auto const jsonData = SerializeUGCIndex(m_UGCIndexes);
+  auto const indexFilePath = GetIndexFilePath();
+  try
+  {
+    FileWriter w(indexFilePath);
+    w.Write(jsonData.c_str(), jsonData.length());
+  }
+  catch (RootException const & exception)
+  {
+    LOG(LWARNING, ("Exception while writing file:", indexFilePath));
```

Comment:
Please, write exception.What(). Also, please, consider to increase severity to LERROR.